app/src/main/java/com/libs/facerecognition/FacePreProcessor.java

```diff
@@ -65,176 +45,107 @@ public FacePreProcessor(Context context){
             try {
-                Thread.sleep(10);
+                Thread.sleep(5);
             } catch (InterruptedException e) {
                 e.printStackTrace();
             }
         }
-        List<Face> faces = result.getResult();
-        for (Face face : faces){
-            android.graphics.Rect rectangle = face.getBoundingBox();
-            facesRect.add(new Rect(rectangle.left,rectangle.top,rectangle.width(),rectangle.height()));
-        }
-        returnResult.fromList(facesRect);
-        return returnResult;
     }
-
     /**
-     * Try to rotate face and trim rest of photo.
+     * Detect face, rotate it and trim.
      *
-     * @param face image of face which will be preprocessed
+     * @param image image of face which will be preprocessed
      * @return Matrix of preprocessed face
      */
-    public Mat preProcessOneFace(Mat face) throws FaceProcessingException {
-        Rect[] eyeArray = findEyesOnImg(face);
-
-        if (eyeArray.length == 2) {
-            face = rotateImageByEye(face, eyeArray);
-        }
-
-        Rect[] faceArray = detectAllFacesUsingML(face).toArray();
-        if (faceArray.length == 0)
+    public Mat preProcessOneFace(Mat image ) throws FaceProcessingException {
+        Task<List<Face>> task = detectAllFacesUsingML(image);
+        waitForTask(task);
+        // Check number of detected faces.
+        List<Face> faces = task.getResult();
+        if (faces.size() == 0)
             throw new FaceProcessingException(FaceProcessingException.NO_FACES);
-        else if (faceArray.length > 1)
+        else if (faces.size() > 1)
             throw new FaceProcessingException(FaceProcessingException.MORE_THAN_ONE_FACE);
 
-        return face.submat(faceArray[0]);
+        // Rotate faces.
+        Face face = faces.get(0);
+        double angle =  -face.getHeadEulerAngleZ();
+        Mat rotatedImage = rotateImageByAngle(image, angle, face.getBoundingBox().centerX(),face.getBoundingBox().centerY());
+        android.graphics.Rect rectangle = face.getBoundingBox();
+
+        // Calculate new bounding box.
+        RotatedRect boundingBox = new RotatedRect(new Point(rectangle.centerX(),rectangle.centerY()),new Size(rectangle.width(),rectangle.height()),angle);
+
+        Log.d(Tag, "Preprocessed one face");
+
+        // Return trimmed face.
+        return rotatedImage.submat(boundingBox.boundingRect());
     }
 
     /**
      * Try to rotate face and trim rest of photo.
      *
-     * @param frame         image with many faces which will be preprocessed
-     * @param detectedFaces Rect of this face
+     * @param frame image with many faces which will be preprocessed
+     * @param listOfFaces List of all faces on this image.
      * @return Matrix of all faces after trimming and rotation
      */
-    public ArrayList<Mat> preProcessAllFaces(Mat frame, MatOfRect detectedFaces) {
+    public ArrayList<Mat> preProcessAllFaces(Mat frame, List<Face> listOfFaces) {
```

Comment:
Don't return ArrayList here, bc you are implying the specific interface. Use List<Mat> instead.